rbac/management/role/v2_model.py

```diff
@@ -0,0 +1,161 @@
+#
+# Copyright 2025 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+
+"""Model for role V2 management."""
+
+import uuid_utils.compat as uuid
+from django.db import models
+from django.db.models import signals
+from django.utils import timezone
+from management.models import Permission, Role
+from management.rbac_fields import AutoDateTimeField
+from rest_framework import serializers
+
+from api.models import TenantAwareModel
+
+
+class RoleV2(TenantAwareModel):
+    """V2 Role model."""
+
+    class Types(models.TextChoices):
+        CUSTOM = "custom"
+        SEEDED = "seeded"
+        PLATFORM = "platform"
+
+    uuid = models.UUIDField(default=uuid.uuid7, editable=False, unique=True, null=False)
+    name = models.CharField(max_length=150, null=False, blank=False)
+    description = models.TextField(null=True, blank=True)
+    type = models.CharField(
+        choices=Types.choices, max_length=20, null=False, blank=False, db_index=True, default=Types.CUSTOM
+    )
+    permissions = models.ManyToManyField(Permission, related_name="v2_roles")
+    children = models.ManyToManyField("self", related_name="parents", symmetrical=False)
+    v1_source = models.ForeignKey(Role, null=True, blank=True, related_name="v2_roles", on_delete=models.SET_NULL)
+    created = models.DateTimeField(default=timezone.now)
+    modified = AutoDateTimeField(default=timezone.now)
+
+    class Meta:
+        ordering = ["name", "modified"]
+        constraints = [
+            models.UniqueConstraint(fields=["name", "tenant"], name="unique role v2 name per tenant"),
+        ]
+
+
+class TypedRoleV2Manager(models.Manager):
+    """Manager for RoleV2 with a specific type."""
+
+    _role_type: str
+
+    def __init__(self, role_type: str):
+        """Initialize the manager."""
+        super().__init__()
+        self._role_type = role_type
+
+    def get_queryset(self):
+        """Get the queryset for the specific type."""
+        return super().get_queryset().filter(type=self._role_type)
+
+
+class TypeValidatedRoleV2Mixin:
+    """Mixin for role types that validates type on init and save."""
+
+    _expected_type: str
```

Comment:
Maybe _expected_type: str = None ?

also we can add into init method if _expected_type was set, in case this mixin is used without setting _expected_type.

Reply:
updated into `_expected_type: str = ""`